### PR TITLE
Add Finetuning Notebook

### PR DESCRIPTION
**Pull request by rawkintrevo**
For easier viewing
https://colab.research.google.com/drive/1Jr4Et2krr_e0ejhvyixxvTM2bik8ONeK?usp=sharing

But remember- trust, but verify. 

TODO:

Add env variable in tests that specify it is a test, if that is set, skip the trainer.train cell

### CODE REVIEW

**Comment by rawkintrevo** [comment]
Now failing bc it can't find a GPU. motion to create a list of exempt recipe notebooks, and put this on the list.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+        "2. Practical: It illustrates how a model's interpretation of domain-specific terms (like 'inheritance') can shift based on the training data.  \n",
+        "  \n",
+        "We'll walk through several key steps:  \n",
+        "- Installing necessary dependencies  \n",
+        "- Loading and exploring the dataset  \n",
+        "- Setting up the quantized model  \n",
+        "- Performing a sanity check  \n",
+        "- Configuring and executing the training process  \n",
+        "  \n",
+        "By the end, we'll have a model that has learned to interpret legal concepts, demonstrating the power and flexibility of transfer learning in NLP.  "
```
Comment (fayvor) [inline recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:37]: Good intro!

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Training Process  \n",
+        "  \n",
+        "With all the preparations complete, we now start the training process. The model will be exposed to numerous examples from our legal dataset, gradually adjusting its understanding of legal concepts.  \n",
+        "  \n",
+        "We'll monitor the training loss over time, which should decrease as the model improves its performance on the task. After training, we'll save the fine-tuned model for future use.  "
```
Comment (fayvor) [inline recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:273]: Which numbers show the training loss?

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+        "  \n",
+        "After fine-tuning and validating our model, a optional step is to make it easily accessible for future use or sharing with the community. The Hugging Face Hub provides an excellent platform for this purpose.  \n",
+        "  \n",
+        "Uploading our model to the Hugging Face Hub offers several benefits:  \n",
+        "1. Easy sharing and collaboration with other researchers or developers  \n",
+        "2. Version control for your model iterations  \n",
+        "3. Integration with various libraries and tools in the Hugging Face ecosystem  \n",
+        "4. Simplified deployment options  \n",
+        "  \n",
+        "We'll demonstrate how to push our fine-tuned model and tokenizer to the Hugging Face Hub, making it available for others to use or for easy integration into other projects. This step is essential for reproducibility and for contributing to the broader NLP community.  "
```
Comment (fayvor) [inline recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:331]: Can we talk about the licensing considerations here? What are the publishing requirements given the data license and the model license?

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+      "metadata": {
+        "id": "JkL-yHtth3Kc"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Loading the Fine-tuned Model  \n",
```
Comment (fayvor) [inline recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:354]: Would it make sense to move this section to another recipe and link this one to it? Since we have the model in-hand already.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Evaluation  \n",
+        "  \n",
+        "Finally, we'll evaluate our fine-tuned model by presenting it with the same 'inheritance' prompt we used in the sanity check. This comparison will reveal how the model's understanding has shifted from a programming context to a legal one.  \n",
+        "  \n",
+        "This step demonstrates the power of transfer learning and domain-specific fine-tuning in natural language processing, showing how we can adapt a general-purpose language model to specialized tasks.  "
```
Comment (fayvor) [inline recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:418]: This is a great experiment; it provides a strong motivation for finetuning.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+        "  \n",
+        "Throughout this notebook, we've been tracking the time taken for various stages of our process. These execution times provide valuable insights into the computational requirements of fine-tuning a large language model.  \n",
+        "  \n",
+        "We'll summarize the time taken for:  \n",
+        "1. Loading the initial model  \n",
+        "2. Performing the sanity check  \n",
+        "3. Setting up the training environment  \n",
+        "4. The actual training process  \n",
+        "  \n",
+        "Understanding these metrics is crucial for resource planning in machine learning projects. It helps in estimating the time and computational power needed for similar tasks in the future, and can guide decisions about hardware requirements or potential optimizations.  \n",
```
Comment (fayvor) [inline recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:466]: Can we say something short about what these particular numbers tell us? Eg., the example metrics taken from running on a T4. Can we extrapolate by model size, data size, and/or GPU size?

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 2 above)
Comment (fayvor) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:273]: I see them now; they appear after the first epoch. Maybe mention that here.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 3 above)
Comment (rawkintrevo) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:331]: The dataset is ASFv2 and we're not publishing, so I'm 99.999% sure it's safe.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 6 above)
Comment (rawkintrevo) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:466]: Candidly, this is just a fluff phrase. I put the timing in because adam was worried about how long it would take.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 4 above)
Comment (rawkintrevo) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:354]: Yea... This was for adam who was worried about giving the workshop live and it taking an hour+ to train the model. I put this little backdoor in so he could get everyone to the point where training had started, then tell them, "OK, now restart your session, we'll just load the finetuned model and play with it now".

I'm open to other ways to skin the cat.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 2 above)
Comment (rawkintrevo) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:273]: It's a fair call out- like i said on the call, I'm trying to avoid going down rabbit holes- if you do you end up with a monstrous notebook, that still isn't high enough quality to use for real fine-tuning, but is very intimidating. I think I'll turn up how often it reports the loss in the logger and add one text block about over fitting and early stopping and then link out to more resources.

**Comment by rawkintrevo** [comment]
Thanks for the review @fayvor!

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 6 above)
Comment (rawkintrevo) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:466]: To clarify- I can, or I can drop all the timing stuff. I could go either way with it.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 4 above)
Comment (fayvor) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:354]: I see. Makes sense.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 2 above)
Comment (fayvor) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:273]: Yeah, I get that you don't want to explain the loss function here. I was really just asking for clarity on where I can see the actual output of the loss numbers (since the recipe says we will "monitor them over time"), b/c they didn't show up right away. Maybe just mention that the first loss numbers will show up after the first epoch is trained.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 3 above)
Comment (fayvor) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:331]: Ok, should we remove the "making it available for others to use" language? It seems we should at least alert users to copyright concerns otherwise.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 6 above)
Comment (fayvor) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:466]: I think we could just tone down the language a bit. Otherwise, if we say that "Understanding these metrics is crucial for resource planning", then we should probably explain the metrics, or at least provide a link to where someone could learn to interpret them and do the resource planning.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 3 above)
Comment (rawkintrevo) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:331]: Yea, I can drop the 'making it available for others to use' language, nbd. But can you expand on copyright concerns? I'm asking in good faith, I truly don't see any- it's an ASFv2 licensed model fine tuned on ASFv2 licensed data.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 2 above)
Comment (rawkintrevo) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:273]: That's a good note. Ill update it to spit out more training info- I also see how I misread your initial comment.

**Comment by fayvor** [comment]
I can approve with some of the minor wording adjustments above.

**Comment by rawkintrevo** [comment]
Ack, I'll have an updated version pushed early this afternoon.

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb (hunk 3 above)
Comment (fayvor) [reply recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:331]: Sorry, I mean bring up copyright concerns for when someone wants to apply this pattern to a different dataset. You could say that that "When considering publishing a fine-tuned model, models and datasets should be evaluated for copyright constraints. Here we are using an ASFv2 licensed model fine-tuned on ASFv2 licensed data." or something similar.

**Comment by fayvor** [comment]
lgtm!

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb [hunk 7]
```diff
@@ -463,7 +463,9 @@
         "3. Setting up the training environment  \n",
         "4. The actual training process  \n",
         "  \n",
-        "Understanding these metrics is crucial for resource planning in machine learning projects. It helps in estimating the time and computational power needed for similar tasks in the future, and can guide decisions about hardware requirements or potential optimizations.  \n",
+        "Understanding these metrics is can be helpful for resource planning in machine learning projects. It helps in estimating the time and computational power needed for similar tasks in the future, and can guide decisions about hardware requirements or potential optimizations.  \n",
+        "\n",
+        "This topic is deep and nuanced, but this can give you an idea of how long your finetuning took on this particular hardware.\n",
```
Comment (fayvor) [inline recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:468]: 👍

File: recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb [hunk 8]
```diff
@@ -328,7 +326,9 @@
         "3. Integration with various libraries and tools in the Hugging Face ecosystem  \n",
         "4. Simplified deployment options  \n",
         "  \n",
-        "We'll demonstrate how to push our fine-tuned model and tokenizer to the Hugging Face Hub, making it available for others to use or for easy integration into other projects. This step is essential for reproducibility and for contributing to the broader NLP community.  "
+        "We'll demonstrate how to push our fine-tuned model and tokenizer to the Hugging Face Hub, making it available for others to use or for easy integration into other projects. This step is essential for reproducibility and for contributing to the broader NLP community.  \n",
+        "\n",
+        "**NOTE:** Check with your own legal counsel before pushing models to Huggingface Hub."
```
Comment (fayvor) [inline recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb:331]: 👍

**Review by fayvor — APPROVED** [review]
